designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,940 @@
+
+#### Overriding Configuration Based on File Patterns
+
+Whereas `.eslintrc` had an `overrides` key that made a hierarchical structure, the `eslint.config.js` file does not have any such hierarchy. Instead, users can return an array of configs that should be used. For example, consider this `.eslintrc` config:
+
+```yaml
+plugins: ["react"]
+rules:
+    react/jsx-uses-react: error
+    semi: error
+
+overrides:
+    - files: "*.md"
+      plugins: ["markdown"],
+      processor: "markdown/markdown"
+```
+
+This can be written in `eslint.config.js` as an array of two configs:
+
+```js
+module.exports = [
+    {
+        files: "*.js",
+        plugins: {
+            react: require("eslint-plugin-react"),
+        },
+        rules: {
+            "react/jsx-uses-react": "error",
+            semi: "error"
+        }
+    },
+    {
+        files: "*.md",
+        processor: require("eslint-plugin-markdown").processors.markdown
+    }
+];
+```
+
+When ESLint uses this config, it will check each `files` pattern to determine which configs apply. Any config with a `files` pattern matching the file to lint will be extracted and used (if multiple configs match, then those configs are merged to determine the final config to use). In this way, returning an array acts exactly the same as the array in `overrides`.
+
+#### Using AND patterns for files
+
+If any entry in the `files` key is an array, then all of the patterns must match in order for a filename to be considered a match. For example:
+
+```js
+module.exports = [
+    {
+        files: [ ["*.test.*", "*.js"] ],
+        rules: {
+            semi: ["error", "always"]
+        }
+    }
+];
+```
+
+Here, the `files` key specifies two glob patterns. The filename `foo.test.js` would match because it matches both patterns whereas the filename `foo.js` would not match because it only matches one of the glob patterns.
+
+**Note:** This feature is primarily intended for backwards compatibility with eslintrc's ability to specify `extends` in an `overrides` block.
+
+#### Ignoring files
+
+With `eslint.config.js`, there are three ways that files and directories can be ignored:
+
+1. **Defaults** - by default, ESLint will ignore `node_modules` and `.git` directories only. This is different from the current behavior where ESLint ignores `node_modules` and all files/directories beginning with a dot (`.`).
+2. **.eslintignore** - the regular ESLint ignore file.
+3. **eslint.config.js** - patterns specified in `ignores` keys when `files` is not specified. (See details below.) 
+
+Anytime `ignores` appears in a config object without `files`, then the `ignores` patterns acts like the `ignorePatterns` key in `.eslintrc` in that the patterns are excluded from all searches before any other matching is done. For example:
```

Comment:
Low-priority idea: it might be helpful to warn the user if there are any other keys in a config object that has `ignores` but no `files` since the other keys won't do anything.